src/codegen/mod.rs

```diff
@@ -645,6 +645,14 @@ impl RetrieveType for Expression {
 }
 
 impl Expression {
+    /// This function loads a reference if necessary
+    pub(crate) fn load(self) -> Expression {
```

Comment:
While I endorse this change, this could probably be used in a lot more places? Given that is true, for the sake of this bugfix, I'd not include it here, but rather do a follow up PR using it everywhere possible.

Reply:
@xermicus I think you are very right! All unary operators can be simple function on `cfg::Expression` with no arguments. We could have a `fn neg_loc(&self, loc: pt::Loc) -> Expression` for a version with location. Binary operators like add be like `fn add(&self, other: Expression) -> Expression` or `fn add_checked(&self, other: Expression)`. These functions can `assert!()` that the arguments are of the same `Type`. We need more functions for variables and literals etc. I think it's a great idea that would make the CFG construction builder interface much, much better.

@LucasSte having said all that, I do have some review comments on this function. I would expect `fn load()` to:
- Always load. How about calling this function `fn load_if_needed()` (can't think of a better name right now)
- Load from storage or memory

Reply:
> While I endorse this change, this could probably be used in a lot more places? Given that is true, for the sake of this bugfix, I'd not include it here, but rather do a follow up PR using it everywhere possible.

I can double check the right places to use this function on this PR already.

Reply:
> @LucasSte having said all that, I do have some review comments on this function. I would expect `fn load()` to:
> 
> * Always load. How about calling this function `fn load_if_needed()` (can't think of a better name right now)
> * Load from storage or memory

I don't think this function should load both from storage and memory, because they are different operations. Load from storage requires an instruction `Instr::LoadStorage`, while the load from memory does not need it. I can clarify this in the function name.

Reply:
Ok fine, but Sean also made some good points that should be addressed in a later PR anyways